internal/session/session.go

```diff
@@ -164,44 +171,25 @@ func (s *Session) ID() uint64 {
 
 // ReceiveFrom receives incoming blocks from the given peer.
 func (s *Session) ReceiveFrom(from peer.ID, ks []cid.Cid, haves []cid.Cid, dontHaves []cid.Cid) {
+	// The SessionManager tells each Session about all keys that it may be
+	// interested in. Here the Session filters the keys to the ones that this
+	// particular Session is interested in.
 	interestedRes := s.sim.FilterSessionInterested(s.id, ks, haves, dontHaves)
 	ks = interestedRes[0]
 	haves = interestedRes[1]
 	dontHaves = interestedRes[2]
 	// s.logReceiveFrom(from, ks, haves, dontHaves)
 
-	// Add any newly discovered peers that have blocks we're interested in to
-	// the peer set
-	isNewPeer := s.sprm.ReceiveFrom(from, ks, haves)
-
-	// Record response timing only if the blocks came from the network
-	// (blocks can also be received from the local node)
-	if len(ks) > 0 && from != "" {
-		s.sprm.RecordPeerResponse(from, ks)
-	}
-
-	// Update want potential
-	s.sws.Update(from, ks, haves, dontHaves, isNewPeer)
+	// Inform the session want sender that a message has been received
+	s.sws.Update(from, ks, haves, dontHaves)
 
 	if len(ks) == 0 {
 		return
 	}
 
```

Comment:
I moved the chunk of code below to be inside the run loop so we can get rid of locking